File: backend/app/api/v1/project.py
from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends
from pydantic import BaseModel
from supabase import Client

from app.dependencies import CurrentUser, get_db
from app.exceptions import NotFoundError, ForbiddenError

router = APIRouter(prefix="/project", tags=["project"])


class CreateProjectRequest(BaseModel):
    name: str
    description: str = ""


class UpdateProjectRequest(BaseModel):
    name: str | None = None
    description: str | None = None
    status: str | None = None


@router.post("")
def create_project(
    body: CreateProjectRequest,
    current_user: CurrentUser = None,
    db: Client = Depends(get_db),
):
    user_id: str = current_user["sub"]
    result = db.table("projects").insert({
        "id": str(uuid.uuid4()),
        "user_id": user_id,
        "name": body.name,
        "description": body.description,
        "status": "draft",
    }).execute()
    return result.data[0]


@router.get("")
def list_projects(
    current_user: CurrentUser = None,
    db: Client = Depends(get_db),
):
    user_id: str = current_user["sub"]
    result = (
        db.table("projects")
        .select("*")
        .eq("user_id", user_id)
        .order("updated_at", desc=True)
        .execute()
    )
    return result.data


@router.get("/{project_id}")
def get_project(
    project_id: str,
    current_user: CurrentUser = None,
    db: Client = Depends(get_db),
):
    user_id: str = current_user["sub"]
    result = (
        db.table("projects")
        .select("""
            *,
            uploads(id, public_url, file_name, upload_status),
            garment_specs(id, spec_json, garment_type, silhouette, is_current, version),
            renders(id, view_type, public_url, render_status),
            techpacks(id, pdf_url, pack_status),
            manufacturing_docs(id, doc_status),
            trend_reports(id, report_status),
            patterns(id, piece_name, pattern_status)
        """)
        .eq("id", project_id)
        .single()
        .execute()
    )
    if not result.data:
        raise NotFoundError("Project", project_id)
    if result.data["user_id"] != user_id:
        raise ForbiddenError()
    return result.data


@router.patch("/{project_id}")
def update_project(
    project_id: str,
    body: UpdateProjectRequest,
    current_user: CurrentUser = None,
    db: Client = Depends(get_db),
):
    user_id: str = current_user["sub"]
    updates = body.model_dump(exclude_none=True)
    if not updates:
        raise NotFoundError("Project", project_id)
    result = (
        db.table("projects")
        .update(updates)
        .eq("id", project_id)
        .eq("user_id", user_id)
        .execute()
    )
    if not result.data:
        raise NotFoundError("Project", project_id)
    return result.data[0]


@router.delete("/{project_id}")
def delete_project(
    project_id: str,
    current_user: CurrentUser = None,
    db: Client = Depends(get_db),
):
    user_id: str = current_user["sub"]
    db.table("projects").delete().eq("id", project_id).eq("user_id", user_id).execute()
    return {"deleted": project_id}

File: frontend/src/lib/types/project.ts
export type ProjectStatus =
  | "draft"
  | "uploading"
  | "analyzing"
  | "spec_ready"
  | "rendering"
  | "render_ready"
  | "techpack_ready"
  | "complete"
  | "error";

export interface Project {
  id: string;
  user_id: string;
  name: string;
  description?: string;
  status: ProjectStatus;
  garment_type?: string;
  thumbnail_url?: string;
  created_at: string;
  updated_at: string;
}

export interface Upload {
  id: string;
  project_id: string;
  public_url: string;
  file_name: string;
  upload_status: string;
}

export interface Render {
  id: string;
  project_id: string;
  spec_id: string;
  view_type: "front" | "back" | "side_left" | "studio_3q";
  public_url?: string;
  render_status: "queued" | "generating" | "complete" | "error";
}

export interface TechPack {
  id: string;
  project_id: string;
  pdf_url?: string;
  pack_status: string;
}

export interface ProjectFull extends Project {
  uploads?: Upload[];
  garment_specs?: Array<{ id: string; spec_json: Record<string, unknown>; is_current: boolean; version: number }>;
  renders?: Render[];
  techpacks?: TechPack[];
  manufacturing_docs?: Array<{ id: string; doc_status: string }>;
  trend_reports?: Array<{ id: string; report_status: string }>;
  patterns?: Array<{ id: string; piece_name: string; pattern_status: string }>;
}
